flowmachine/flowmachine/core/server/query_schemas/meaningful_locations.py

```diff
@@ -0,0 +1,311 @@
+# This Source Code Form is subject to the terms of the Mozilla Public
+# License, v. 2.0. If a copy of the MPL was not distributed with this
+# file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+from marshmallow import Schema, fields, post_load
+from typing import Union, Dict, List
+
+from flowmachine.features import (
+    MeaningfulLocations,
+    MeaningfulLocationsOD,
+    MeaningfulLocationsAggregate,
+    HartiganCluster,
+    CallDays,
+    EventScore,
+    subscriber_locations,
+)
+from .base_exposed_query import BaseExposedQuery
+from .custom_fields import (
+    AggregationUnit,
+    SubscriberSubset,
+    TowerHourOfDayScores,
+    TowerDayOfWeekScores,
+)
+
+__all__ = [
+    "MeaningfulLocationsAggregateSchema",
+    "MeaningfulLocationsAggregateExposed",
+    "MeaningfulLocationsBetweenLabelODMatrixSchema",
+    "MeaningfulLocationsBetweenLabelODMatrixExposed",
+]
+
+
+class MeaningfulLocationsAggregateSchema(Schema):
+    start_date = fields.Date(required=True)
+    stop_date = fields.Date(required=True)
+    aggregation_unit = AggregationUnit(required=True)
+    label = fields.String(required=True)
+    labels = fields.Dict(
+        required=True, keys=fields.String(), values=fields.Dict()
+    )  # TODO: use custom field here for stricter validation!
+    tower_hour_of_day_scores = TowerHourOfDayScores(required=True)
+    tower_day_of_week_scores = TowerDayOfWeekScores(required=True)
+    tower_cluster_radius = fields.Float(required=False, default=1.0)
+    tower_cluster_call_threshold = fields.Integer(required=False, default=0)
+    subscriber_subset = SubscriberSubset(required=False)
+
+    @post_load
+    def make_query_object(self, params):
+        return MeaningfulLocationsAggregateExposed(**params)
+
+
+def _make_meaningful_locations_object(
```

Comment:
One main reason this wasn't flattened in my original implementation is that 99% of the time, you want (like with the aggregates) to put the queries it composes into cache before starting to run, to avoid paying a massive performance penalty. Obviously the signature is nicer this way, but I do have a concern that this puts a barrier in the way of efficiently caching stuff in order to make a primarily machine facing interface more readable to humans.

Reply:
That's a good point, although I feel the user should not need to worry about this. My hunch is that we can actually pretty easily achieve this by adding marshmallow schemas for these sub-queries which we simply don't expose via the API. Might need a little re-rigging of the marshmallow machinery but it should actually be fairly easy (I hope).